### PR TITLE
[Improve][KafkaSource] Move Kafka Source Create By Source Factory

### PR DESCRIPTION
**Pull request by Hisoka-X**
## Purpose of this pull request

Kafka Source use SourceFactory to create instance, so user can implement theirs own catalog, then implement multi table in kafka.

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/serialization/DeserializationSchemaWithTopic.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    T deserialize(String topic, byte[] message) throws IOException;
+
+    default void deserialize(String topic, byte[] message, Collector<T> out) throws IOException {
+        try {
+            T deserialize = deserialize(topic, message);
+            if (deserialize != null) {
+                out.collect(deserialize);
+            }
+        } catch (IOException e) {
+            throw new IOException(e);
```
Comment (EricJoy2048) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/serialization/DeserializationSchemaWithTopic.java:43]: Is it a better choice to specify an Exception for SeaTunnel?

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java [hunk 2]
```diff
@@ -170,6 +170,22 @@ public static List<CatalogTable> getCatalogTables(
         return catalogTables;
     }
 
+    public static CatalogTableUtil buildWithReadonlyConfig(ReadonlyConfig option) {
+        if (option.get(FIELDS) == null) {
```
Comment (EricJoy2048) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java:174]: SeaTunnelException? or SeaTunnelCatalogException?

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/config/Config.java [hunk 3]
```diff
@@ -84,25 +86,18 @@ public class Config {
...
-                    .objectType(Config.class)
-                    .noDefaultValue()
-                    .withDescription(
-                            "The structure of the data, including field names and field types.");
-
     public static final Option<MessageFormat> FORMAT =
             Options.key("format")
                     .enumType(MessageFormat.class)
-                    .defaultValue(MessageFormat.JSON)
+                    .defaultValue(MessageFormat.TEXT)
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/config/Config.java:92]: Will modifying the default values not have compatibility issues?

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/config/Config.java [hunk 4]
```diff
@@ -84,25 +86,18 @@ public class Config {
...
                     .withDescription(
-                            "Data format. The default format is json. Optional text format. The default field separator is \", \". "
+                            "Data format. The default format is text. Optional json format. The default field separator is \", \". "
                                     + "If you customize the delimiter, add the \"field_delimiter\" option.");
 
     public static final Option<String> FIELD_DELIMITER =
             Options.key("field_delimiter")
                     .stringType()
-                    .noDefaultValue()
+                    .defaultValue(DEFAULT_FIELD_DELIMITER)
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/config/Config.java:100]: Same as above.

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSource.java [hunk 5]
```diff
@@ -239,44 +204,47 @@ public void setJobContext(JobContext jobContext) {
...
+            ReadonlyConfig option, SeaTunnelDataType<SeaTunnelRow> dataType) {
+        deserializationSchema = getDeserializationSchema(option, dataType);
+    }
+
+    private DeserializationSchema<SeaTunnelRow> getDeserializationSchema(
+            ReadonlyConfig option, SeaTunnelDataType<SeaTunnelRow> typeInfo) {
+        MessageFormat format = option.get(FORMAT);
+        switch (format) {
+            case JSON:
+                if (typeInfo instanceof SeaTunnelRowType) {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSource.java:217]: If SeaTunnelRowType is needed, why not update the method to `private DeserializationSchema<SeaTunnelRow> getDeserializationSchema(
            ReadonlyConfig option, SeaTunnelRowType typeInfo)` ?

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceFactory.java [hunk 6]
```diff
@@ -53,6 +66,27 @@ public OptionRule optionRule() {
...
+            } else {
+                Map<String, SeaTunnelRowType> rowTypeMap = new HashMap<>();
+                for (CatalogTable catalogTable : context.getCatalogTables()) {
+                    String tableId = catalogTable.getTableId().toTablePath().toString();
+                    rowTypeMap.put(tableId, catalogTable.getTableSchema().toPhysicalRowDataType());
+                }
+                dataType = new MultipleRowType(rowTypeMap);
+            }
+            return (SeaTunnelSource<T, SplitT, StateT>)
+                    new KafkaSource(context.getOptions(), dataType);
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/source/KafkaSourceFactory.java:86]: Why not set CatalogTableList to Source?

**Comment by EricJoy2048** [comment]
Please update the document reference https://github.com/apache/incubator-seatunnel/pull/4590/files

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java [hunk 7]
```diff
@@ -170,6 +170,22 @@ public static List<CatalogTable> getCatalogTables(
...
+            throw new RuntimeException(
+                    "Schema config need option [schema], please correct your config first");
+        }
+        TableSchema tableSchema = parseTableSchema(option.get(FIELDS));
+        return new CatalogTableUtil(
+                CatalogTable.of(
+                        // TODO: other table info
+                        TableIdentifier.of("", "", ""),
```
Comment (hailin0) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/table/catalog/CatalogTableUtil.java:182]: Is there no need to define an table identifier?

File: seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/config/Config.java (hunk 3 above)
Comment (hailin0) [reply seatunnel-connectors-v2/connector-kafka/src/main/java/org/apache/seatunnel/connectors/seatunnel/kafka/config/Config.java:92]: +1

**Comment by Hisoka-X** [comment]
Hi, before the Spark/Flink translation support Factory SPI to create source/sink. The PR will not update. I will modify it after Spark/Flink support Factory SPI

**Pull request closed without merging**